frontend/src/metabase/query_builder/QueryVisualization.jsx

```diff
@@ -206,168 +205,87 @@ export default class QueryVisualization extends Component {
-                    // "display", "visualization_settings", etc, (to ensure the correct visualization is shown)
-                    // BUT the last executed "dataset_query" (to ensure data matches the query)
-                    let card = {
-                        ...this.props.card,
-                        dataset_query: this.state.lastRunDatasetQuery
-                    };
-                    viz = (
-                        <Visualization
-                            className="full"
-                            series={[{ card: card, data: this.props.result.data }]}
-                            isEditing={true}
-                            // Table:
-                            setSortFn={this.props.setSortFn}
-                            cellIsClickableFn={this.props.cellIsClickableFn}
-                            cellClickedFn={this.props.cellClickedFn}
-                            onUpdateVisualizationSetting={this.props.onUpdateVisualizationSetting}
-                            onUpdateVisualizationSettings={this.props.onUpdateVisualizationSettings}
-                        />
-                    );
-                }
+            if (error) {
+                viz = <VisualizationError error={error} card={card} duration={result.duration} />
+            } else if (result.data) {
+                viz = <VisualizationResult lastRunDatasetQuery={this.state.lastRunDatasetQuery} {...this.props} />
             }
         }
 
-        var wrapperClasses = cx('wrapper full relative mb2 z1', {
-            'flex': !this.props.isObjectDetail,
-            'flex-column': !this.props.isObjectDetail
+        const wrapperClasses = cx('wrapper full relative mb2 z1', {
+            'flex': !isObjectDetail,
+            'flex-column': !isObjectDetail
         });
 
-        var visualizationClasses = cx('flex flex-full Visualization z1 px1', {
-            'Visualization--errors': (this.props.result && this.props.result.error),
-            'Visualization--loading': this.props.isRunning
+        const visualizationClasses = cx('flex flex-full Visualization z1 px1', {
+            'Visualization--errors': (result && result.error),
+            'Visualization--loading': isRunning
         });
 
         return (
             <div className={wrapperClasses}>
                 {this.renderHeader()}
-                {loading}
+                { isRunning && (
+                    <div className="Loading spread flex flex-column layout-centered text-brand z2">
+                        <LoadingSpinner />
+                        <h2 className="Loading-message text-brand text-uppercase my3">Doing science...</h2>
+                    </div>
+                )}
                 <div className={visualizationClasses}>
                     {viz}
                 </div>
             </div>
         );
     }
 }
+
+
+const VisualizationResult = ({card, isObjectDetail, lastRunDatasetQuery, result, ...rest}) => {
```

Comment:
Mostly just for the convenience of working on it while refactoring. Planning on moving them out into their own files.